src/common/service/NetRemoteDataStreamingReactors.cxx

```diff
@@ -107,19 +123,33 @@ DataStreamWriter::OnWriteDone(bool isOk)
 void
 DataStreamWriter::OnCancel()
 {
+    LOGD << "Enter OnCancel";
+
     // The RPC is cancelled by the client, so call Finish to complete it from the server perspective.
-    Finish(grpc::Status::CANCELLED);
+    if (!m_isCancelled.load(std::memory_order_relaxed)) {
+        m_isCancelled.store(true, std::memory_order_relaxed);
```

Comment:
If you're just trying to set `m_isCancelled` only if it's not already set, then this should use `compare_exchange_strong` instead to avoid a race where `m_isCancelled` gets set in between the `load` and `strore`:

```suggestion
    const bool isCanceledExpected{ false };
    if (m_isCancelled.compare_exchange_strong(isCanceledExpected, true, std::memory_order_relaxed, std::memory_order_relaxed)) { 
```

Reply:
Technically `m_isCancelled` should only be set here in `OnCancel`, so perhaps I can simply do `m_isCancelled.store(true, std::memory_order_relaxed);` without the check?

Reply:
I don't think so, because you only want to call `Finish()` once, right?